### ISSUE TITLE
Provide information on whether a value was defaulted

### ISSUE THREAD

**Issue by csaftoiu** [docopt/docopt#341]
Consider:

```
"""Usage:
    foo.py [--knob=<n>]

Options:
    --knob=<n>   Set the knob to n. [default: 14]
"""
import pprint

from docopt import docopt


pprint.pprint(docopt(__doc__))
```

The following two command line invocations give the same result:

    $ python foo.py
    {'--knob': '14'}
    $ python foo.py --knob 14
    {'--knob': '14'}

This means there is no way to tell right now whether a parameter was explicitly passed by the user, or if it was not passed and a default was used.

I propose adding a `'[defaulted]'` key to the returned dict, which contains a set of all the values whose defaults were used. The new result would be:

    $ python foo.py
    {'--knob': '14', '[defaulted]': set(['--knob'])}
    $ python foo.py --knob 14
    {'--knob': '14', '[defaulted]': set([])}

The rationale is this will make it much easier to handle config files where you always want the command-line option to trump the config file option. As it is now, it's impossible to do cleanly using the docopt API.

**Comment by joell** [comment]
As a further motivating example for resolving this ambiguity, consider merging command-line options with configuration file data loaded into a compatible dict object.  Without any defaults specified in the docopt text, this is easy: Just merge the cli options and configuration file dictionaries by using the configuration-file value for every option that docopt assigned `None` to.

However, once default options are introduced, there's a problem.  If the user passed an explicit option with the same value as the default, then the user's value should be used, not the one in the configuration file.  But if the user did *not* specify a value and the configuration file holds a different value than the default, the configuration file's value should override. Because it is impossible to determine if the option's value was set by default or not, this cannot be resolved.

**Comment by SeeSpotRun** [comment]
If there is a config file, what is the meaning of the cli default value?

Consider running the above example when the config file has a value of 12:
```
my.conf:
    <snip>
    knob=12
    <snip>
```

If the user runs `$ python foo.py --knob 14` then docopt will return a value of 14 and merging cli with conf will work just fine.
If the user runs `$ python foo.py` then the desired outcome is knob=12 (the conf file value).  So what is the point of specifying a default value in `__doc__`?  It's actually misleading because the Usage string implies that omitting `--knob` will default to 14 when in fact it defaults to the config file value.  The only way to correctly inform the user of the default values is to write the config file values into `__doc__`.  Actually that's not hard to do:
```
"""Usage:
    foo.py [--knob=<n>]

Options:
    --knob=<n>   Set the knob to n. [default: {knob}]
"""

import pprint

from docopt import docopt

config = read_config_into_dict()

pprint.pprint(docopt(__doc__.format(**config)))

```
)

**Comment by csaftoiu** [comment]
cli default value makes sense semantically. It means: "If no value is specified anywhere, this is the value used." If the program can run meaningfully with a default value, then it should. With your approach the program wouldn't be able to run unless there exists a config file with a value for the given parameter. But it can make sense where a user can download a program and just run it without needing to set up any config file.

To recap, the semantics are sensible:
* If the value is not specified in either command line or config file, the default value is used.
* If the value is specified in the config file, then the config value is used. 
* If the value is specified on the command line, the command line value is used. 
* If the value is specified in both, the command line value should be used. 

This last case is impossible to implement with docopt as it is now.

**Comment by joell** [comment]
I was thinking of the case in which the config file itself is optional and can be user-provided.  In that case, if no configuration file were provided by the user, then the defaults given in the docopt description would prevail.  If the user provides a configuration file, then the config-file values would pervail over the docopt description.

For a real-world example of a daemon that has a configuration file that mirrors its cli options and works like this, take a look at [dnsmasq](https://linux.die.net/man/8/dnsmasq).  It allows for arbitrary configuration files via its `--conf-file` command-line option, but it also specifies defaults for some of its cli options (e.g., `--cache-size`).

**Comment by SeeSpotRun** [comment]
Ok that makes sense.  So the code would be something like:
```
"""Usage:
    foo.py [--knob=<n>]

Options:
    --knob=<n>   Set the knob to n. [default: 14]
"""

from docopt import docopt

config = read_config_into_dict()

options = docopt(__doc__)   # (note: docopt modified to return '[defaulted]' set)

knob = options['--knob'] if ('--knob' not in options['[defaulted]'] or 'knob' not in config) else config['knob']
```

You could still do this quite easily without any changes to docopt:
```
"""Usage:
    foo.py [--knob=<n>]

Options:
    --knob=<n>   Set the knob to n. (default={knob})
"""

defaults = {'knob': '14'}

from docopt import docopt

config = read_config_into_dict()

options = docopt(__doc__.format(**defaults))

knob = options['--knob'] or config['knob'] or defaults['knob']
```

**Comment by joell** [comment]
Yes, an approach like that will attain the functionality needed, and in fact I adopted something similar when I faced this.

However, the same case could be made that the feature of having defaults *at all* in docopt is unnecessary because this approach can easily implement it, right?  But the whole point of adding the feature of allowing for default values to be specified was to save developers the work of reinventing that wheel.

That aside, the one remaining detraction from this approach is that it requires a different syntax.  If you have the need to identify whether a given option's value was explicitly provided or defaulted, then you are forced to change from the "[default: value]" standard docopt syntax to some non-standard "(default: value)" syntax.  If you have a suite of programs that you all want to have a similar look and feel in their command-line interfaces, then for consistency's sake you would have to *also* use the "(default: value)" syntax for the rest those programs, even if they may not share the same need and could have instead saved time and SLoC by using the built-in "[default: value]" syntax.  At that point, you are now re-implementing the entire default-value wheel regardless.

So if docopt is going to take the position that default values are a feature it wants to provide in the first place, then I would argue that the caller being able to determine whether a given value was either explicitly set or defaulted should be a requirement.

If it's just a matter of developer time here, just let me know and I'd be happy to write this up and submit a PR.

**Comment by SeeSpotRun** [comment]
@joell it seems developer time is very scarce (https://github.com/docopt/docopt/issues/371#issuecomment-295928562) so a PR is a good idea.  But also maybe smart to check with @keleshev whether he supports the concept first since he is the only one who can accept PR's.

**Comment by csaftoiu** [comment]
I've put it on my "to do someday list" to implement this and make a PR... this may be a matter of months until I get around to it. But it would be good to know if @keleshev supports it before I do that :)

**Comment by gstaniak** [comment]
Bump up. Found this discussion after hitting exactly the same issue -- admittedly, a cosmetic one, and thanks for the "dictionary of defaults" hint, but something in me cringes every time I write code in which there's more than one place where I need to maintain some config values.
